### PR TITLE
Recursive transactions implemented.

### PR DESCRIPTION
**Pull request by spiresheep**

### CODE REVIEW

File: source/array_table_model.ts [hunk 1]
```diff
@@ -20,11 +20,15 @@ export class ArrayTableModel extends TableModel {
    *  the parent transaction.
    */
   public beginTransaction(): void {
+    this.transactionCount = this.transactionCount + 1;
```
Comment (kranar) [inline source/array_table_model.ts]: Use ```++x```

File: tests/array_table_model_tester.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { AddRowOperation, ArrayTableModel, MoveRowOperation, Operation,
 export class ArrayTableModelTester {
 
   /** Tests adding rows. */
-  @Test()
+//  @Test()
```
Comment (kranar) [inline tests/array_table_model_tester.ts]: Put back all the unit tests.

File: tests/array_table_model_tester.ts [hunk 3]
```diff
@@ -199,4 +199,37 @@ export class ArrayTableModelTester {
     Expect(destinationIndex).toEqual(2);
     listener.unlisten();
   }
+
+  /** Tests movings rows. */
```
Comment (kranar) [inline tests/array_table_model_tester.ts]: Fix documentation.

File: source/array_table_model.ts [hunk 4]
```diff
@@ -20,11 +20,15 @@ export class ArrayTableModel extends TableModel {
...
   }
 
   /** Ends a transaction. */
   public endTransaction(): void {
-    this.dispatcher.dispatch(this.operations);
+    this.transactionCount = this.transactionCount - 1;
+    if(this.transactionCount === 0) {
+      this.dispatcher.dispatch(this.operations);
```
Comment (kranar) [inline source/array_table_model.ts:32]: This is a bug, create a unit test that performs two transactions in sequence (not recursively) and validate the operations. For example perform an addRow transaction and ensure the operations published is a single ```AddRowOperation```, and then remove that same row and ensure the published operations is a single ```RemoveRowOperation```.

Correct the bug you identify.

**Review by kranar — CHANGES_REQUESTED** [review]

File: source/array_table_model.ts (hunk 4 above)
Comment (spiresheep) [reply source/array_table_model.ts:32]: I'm not seeing what the error is.

Comment (kranar) [reply source/array_table_model.ts:32]: Push your unit test.

File: tests/array_table_model_tester.ts [hunk 5]
```diff
@@ -232,4 +232,26 @@ export class ArrayTableModelTester {
...
+      } else {
+        Expect(false).toEqual(true);
+      }
+    };
+    const listener = model.connect(slot);
+    model.addRow([1, 2, 3]);
+    model.addRow([9, 0, 2]);
+    model.removeRow(1);
+    model.removeRow(0);
+    Expect(numberOfOperations).toEqual(4);
```
Comment (kranar) [inline tests/array_table_model_tester.ts]: On line 254, ```numberOfOperations``` should be equal to 1. In fact, the only way ```numberOfOperations``` can be greater than 1 is if you call ```beginTransaction```, run a series of operations and then call ```endTransaction```. Consider the API of the table model to see why this must be the case.

**Review by kranar — CHANGES_REQUESTED** [review]

File: tests/array_table_model_tester.ts (hunk 5 above)
Comment (spiresheep) [reply tests/array_table_model_tester.ts]: Just to make sure I understand correctly, operations is just a record of the last transaction that happened to the table? Not a list of every transaction that has happened to the table?

Comment (kranar) [reply tests/array_table_model_tester.ts]: If you're referring to ```this.operations```, that's up to you to implement and use as you best see fit. In terms of the API you are correct that there is nothing about the TableModel API that represents a list of every transaction that ever happened. The API only emits a signal at the end of a top-level transaction and the value it emits is a list of every operation performed in that transaction.

**Pull request merged**